### PR TITLE
refactor: dockerfile

### PR DESCRIPTION
**Pull request by AVVS**
Hi @gmr, excellent job with autocluster for 3.6.x :)
Below is a proposal for refactored Dockerfile and repo structure on disk.

Idea was to reduce amount of layers as much as we can, add volume for storing mnesia db, launch script to fix permissions on it during launch. People generally argue that we should use more layers and cache topmost ones, but I personally disagree, since when you build new versions of rabbitmq - you would likely update plugin versions. 

This release is opinionated, but maybe it would give some ideas to make the best out of what you already have and my approach

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,52 +1,57 @@
-FROM alpine:edge
+FROM gliderlabs/alpine:3.4
```
Comment (gmr) [inline Dockerfile]: There is a move from ``gliderlabs/alpine`` to ``alpine`` being  top level official image:

https://hub.docker.com/_/alpine/

File: Dockerfile [hunk 2]
```diff
@@ -1,52 +1,57 @@
...
     rabbitmq_management \
+    rabbitmq_management_visualiser \
+    rabbitmq_consistent_hash_exchange \
+    rabbitmq_federation \
+    rabbitmq_federation_management \
+    rabbitmq_mqtt \
+    rabbitmq_shovel \
+    rabbitmq_shovel_management \
+    rabbitmq_stomp \
+    rabbitmq_web_stomp \
```
Comment (gmr) [inline Dockerfile]: The Dockerfile included is for dev purposes -- I've been maintaining https://github.com/gmr/alpine-rabbitmq-autocluster which may be closer to what you're looking for...

File: Dockerfile [hunk 3]
```diff
@@ -1,52 +1,57 @@
...
+    RABBITMQ_PLUGINS_DIR=/usr/lib/rabbitmq/plugins \
+    RABBITMQ_PLUGINS_EXPAND_DIR=/var/lib/rabbitmq/plugins \
+    LANG=en_US.UTF-8 \
+    # Set this so that CTRL+G works properly
+    TERM=xterm
 
 RUN \
-  apk --update add bash coreutils curl erlang erlang-asn1 erlang-crypto erlang-eldap erlang-erts erlang-inets erlang-mnesia erlang-os-mon erlang-public-key erlang-sasl erlang-ssl erlang-syntax-tools erlang-xmerl xz && \
+  apk --update add \
+    coreutils curl xz "su-exec>=0.2" \
```
Comment (gmr) [inline Dockerfile]: +1 on su-exec... I've been meaning to move to using that.

File: Dockerfile (hunk 1 above)
Comment (AVVS) [inline Dockerfile]: didn't know, thx for info

File: Dockerfile [hunk 4]
```diff
@@ -1,52 +1,57 @@
...
+    rabbitmq_federation \
+    rabbitmq_federation_management \
+    rabbitmq_mqtt \
+    rabbitmq_shovel \
+    rabbitmq_shovel_management \
+    rabbitmq_stomp \
+    rabbitmq_web_stomp \
     autocluster
 
+VOLUME $HOME
```
Comment (gmr) [inline Dockerfile]: My main gripe with ``VOLUME`` these days is the poor behavior of docker-compose leaving volumes behind by default. Otherwise I'm +1 on this.

File: root/usr/lib/etc/rabbitmq/rabbitmq.config [hunk 5]
```diff
@@ -18,5 +18,11 @@
                     {mirroring, info}]},
       {vm_memory_high_watermark, 0.8}
     ]
+  },
+  {
+    rabbitmq_management,
+    [
+      {rates_mode, basic}
```
Comment (gmr) [inline root/usr/lib/etc/rabbitmq/rabbitmq.config]: This is the default behavior: https://www.rabbitmq.com/management.html#fine-stats

File: Dockerfile (hunk 4 above)
Comment (AVVS) [inline Dockerfile]: I think there is a setting to remove that using "docker", but not docker-compose, however, we can add a cleanup script like this:

```sh
docker volume ls -qf dangling=true | xargs -r docker volume rm
docker images --quiet --filter "dangling=true" | xargs -r docker rmi
```

File: root/usr/lib/etc/rabbitmq/rabbitmq.config (hunk 5 above)
Comment (AVVS) [inline root/usr/lib/etc/rabbitmq/rabbitmq.config]: wanted to be more explicit here, so that users who want to change it - would know where to dig

**Comment by gmr** [comment]
Thanks for the improvements!

I'm not a fan of having the binary of the last release in the git repo, please remove those. Since this is primarily the dev repo for the plugin, I don't think it's unreasonable to expect the plugins to be in the ``plugins`` directory when building the docker image. This is done via the ``make dist`` command.

File: Dockerfile (hunk 1 above)
Comment (gmr) [inline Dockerfile]: I am +1 for pinning to 3.4 here though.

File: .gitignore [hunk 6]
```diff
@@ -15,11 +15,10 @@ docker/rabbitmq_autocluster*ez
 .vagrant
 .idea
 cover
-plugins
+/plugins
```
Comment (gmr) [inline .gitignore]: my .gitignore foo might be off, but wouldn't that make this the root of the fs and not a subdir of the current directory?

**Comment by gmr** [comment]
Given this is the dev repo for the plugin, I think it makes sense to ``mv root docker`` or ``docker-root`` or something along those lines to make it clear those are docker image specific files.

File: .gitignore (hunk 6 above)
Comment (AVVS) [inline .gitignore]: nope, thats the root of the .git repo

**Comment by AVVS** [comment]
agree with comments regarding plugins since thats a dev repo

**Comment by gmr** [comment]
Thanks again for the changes, I'll merge it if you're able to address the few nits I pointed out.

**Comment by AVVS** [comment]
@gmr should be good for the squash

File: Dockerfile [hunk 7]
```diff
@@ -13,9 +13,7 @@ ENV RABBITMQ_VERSION=3.6.2 \
     RABBITMQ_PID_FILE=/var/lib/rabbitmq/rabbitmq.pid \
     RABBITMQ_PLUGINS_DIR=/usr/lib/rabbitmq/plugins \
     RABBITMQ_PLUGINS_EXPAND_DIR=/var/lib/rabbitmq/plugins \
-    LANG=en_US.UTF-8 \
-    # Set this so that CTRL+G works properly
-    TERM=xterm
```
Comment (gmr) [inline Dockerfile]: Was this removal intentional?

Comment (AVVS) [inline Dockerfile]: yes, TERM is not really needed unless you would open "erl" console, which you likely wouldn't
